src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py

```diff
@@ -0,0 +1,154 @@
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author A. B. G. Chalk, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the tests for the DebugChecksumTrans.'''
+
+from psyclone.psyir.nodes import Routine
+from psyclone.psyir.transformations import DebugChecksumTrans
+from psyclone.tests.utilities import Compile
+
+
+def test_checksum(fortran_reader, fortran_writer, tmpdir):
+    ''' Test the behaviour of the debug_checksum_trans'''
+
+    code = """
+    subroutine test
+        integer, dimension(1:100) :: a, b, c
+        logical, dimension(1:100) :: f
+        integer :: i
+        integer :: d
+
+        do i = 1, 100
+            a(i) = c(i) + d
+            b(i) = 2 * i
+            f(i) = .true.
+        enddo
+    end subroutine
+    """
+
+    psyir = fortran_reader.psyir_from_source(code)
+
+    DebugChecksumTrans().apply(psyir.children[0].children[0])
+
+    out = fortran_writer(psyir)
+    correct = "integer :: PSYCLONE_INTERNAL_line_"
+    assert correct in out
+    correct = """  enddo
+  PSYCLONE_INTERNAL_line_ = __LINE__
+  ! PSyclone DebugChecksumTrans-generated checksums
+  PRINT *, "PSyclone checksums from test at line:", PSYCLONE_INTERNAL_line_ + 1
+  PRINT *, "b checksum", SUM(b(:))
+  PRINT *, "a checksum", SUM(a(:))"""
+    assert correct in out
+    assert Compile(tmpdir).string_compiles(out)
+
+    # Check the checksums are in the right place if there's no hierarchy of
+    # nodes.
+    code = """
+    subroutine test
+        integer, dimension(1:100) :: a, b
+        a(:) = 1
+        b(:) = 2
+    end subroutine
+    """
+    psyir = fortran_reader.psyir_from_source(code)
+
+    DebugChecksumTrans().apply(psyir.children[0].children[:])
+    out = fortran_writer(psyir)
+
+    correct = "integer :: PSYCLONE_INTERNAL_line_"
+    assert correct in out
+    correct = """PSYCLONE_INTERNAL_line_ = __LINE__
```

Comment:
Could you extend this slightly to either include the `b(:) = ...` or the `end subroutine` so that we know the statements are in the right location.

Reply:
Done. Also added a test where we didn't do the entire schedule.